File: app/src/main/java/com/beone/bestpractice/ui/CountryFragment.java
package com.beone.bestpractice.ui;

import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.beone.bestpractice.R;
import com.beone.bestpractice.adapter.CountryAdapter;
import com.beone.bestpractice.adapter.OnClickItemListener;
import com.beone.bestpractice.local.model.CountryEntity;
import com.beone.bestpractice.utility.Resource;

import java.util.List;

public class CountryFragment extends Fragment implements OnClickItemListener {

  private static final String TAG = CountryFragment.class.getName();
  private CountryViewModel mViewModel;

  public static CountryFragment newInstance() {
    return new CountryFragment();
  }

  private RecyclerView recyclerView;
  private CountryAdapter countryAdapter;

  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setRetainInstance(true);
  }

  @Override
  public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                           @Nullable Bundle savedInstanceState) {
    View v = inflater.inflate(R.layout.country_fragment, container, false);
    recyclerView = v.findViewById(R.id.recyclerview);
    initrecycler();

    return v;
  }

  private void initrecycler(){
    countryAdapter = new CountryAdapter(null,requireContext(),this);
    recyclerView.setItemViewCacheSize(30);
    recyclerView.setDrawingCacheEnabled(true);
    recyclerView.setDrawingCacheQuality(View.DRAWING_CACHE_QUALITY_LOW);
    recyclerView.setAdapter(countryAdapter);
  }

  private void subscribeviewmodel(){
    mViewModel.getCountrys().observe(getViewLifecycleOwner(), new Observer<Resource<List<CountryEntity>>>() {
      @Override
      public void onChanged(Resource<List<CountryEntity>> listResource) {
        if (listResource.data != null) {
          switch (listResource.status) {
            case ERROR:
              Log.e(TAG, "onChanged: cannot refresh the cache.");
              Log.e(TAG, "onChanged: ERROR message: " + listResource.message);
              Log.e(TAG, "onChanged: status: ERROR, #state: " + listResource.data.size());
              Toast.makeText(getContext(), listResource.message, Toast.LENGTH_SHORT).show();
            case SUCCESS:
              Log.i(TAG, "onChanged: Success Data Get");
              countryAdapter.updatedata(listResource.data);
              break;
            case LOADING:
              Log.i(TAG, "onChanged: Loading");
              Toast.makeText(getContext(), "Loading...", Toast.LENGTH_SHORT).show();
              break;
          }
        }
      }
    });
  }

  @Override
  public void onActivityCreated(@Nullable Bundle savedInstanceState) {
    super.onActivityCreated(savedInstanceState);
    mViewModel = new ViewModelProvider(this).get(CountryViewModel.class);
    subscribeviewmodel();
  }


  @Override
  public <T> void onClick(T data) {
      mViewModel.test();
  }
}

File: app/src/main/java/com/beone/bestpractice/ui/CountryFragment.kt
package com.beone.bestpractice.ui

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatDelegate
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import com.beone.bestpractice.R
import com.example.core.data.Resource
import com.example.core.domain.model.Country
import com.example.core.ui.CountryMultipleTypeAdapter
import com.example.core.ui.OnClickItemListener
import kotlinx.android.synthetic.main.country_fragment.*
import org.koin.android.viewmodel.ext.android.viewModel


class CountryFragment : Fragment(), OnClickItemListener {
    private val mViewModel: CountryViewModel by viewModel()

    private lateinit var countryMultipleTypeAdapter: CountryMultipleTypeAdapter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val v = inflater.inflate(R.layout.country_fragment, container, false)
        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM)

       countryMultipleTypeAdapter =  CountryMultipleTypeAdapter("Covid 19", requireContext(), this)

        return v
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        tv_favoriteList.setOnClickListener {
            val fragmentTransaction = activity?.supportFragmentManager?.beginTransaction()
            fragmentTransaction?.addToBackStack("favorite")?.replace(R.id.framelayout, FavoriteFragment(), null)
                    ?.commit()
        }
        mViewModel.country.observe(viewLifecycleOwner, Observer {
            if (it != null) {
                when (it) {
                    is Resource.Loading -> Log.d(TAG, "onViewCreated: Loading")
                    is Resource.Success -> {
                        Log.d(TAG, "onViewCreated: Success")
                        countryMultipleTypeAdapter.updatedata(it.data)
                    }
                    is Resource.Error -> {
                        Log.d(TAG, "onViewCreated: Error")
                    }
                }
            }
        })
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        with(recyclerview){
            this?.setHasFixedSize(true)
            this?.adapter = countryMultipleTypeAdapter
         }
    }

    override fun <T> onClick(data: T) {
        val i = Intent(requireContext().applicationContext, Class.forName("com.example.favoritefeature.DetailActivity"))
        val datas: Country? =  data as Country?
        i.putExtra("COUNTRYBUNDLE",datas)
        startActivity(i)
    }

    override fun onDestroyView() {
        recyclerview.adapter = null
        super.onDestroyView()
    }

    companion object {
        private val TAG = CountryFragment::class.java.name
    }
}